### PR TITLE
Extract Build verb from Compile

### PR DESCRIPTION
**Pull request by cdmihai**
Addresses the first proposed PR from issue #532 of separating the existing code into Compile and Build:
- Compile's only responsibility is to know how to invoke the compiler and how to do native compilation
- Build's responsibility is to compile a project graph (root project + dependencies) and delegates to Compile for actual compilation.

I tried to factor out the common code:
- Added extension methods to ProjectContext for resolving the output paths
- Extracted out a common definition for the build and compile command line interface. Unfortunately I could not figure out how to properly share it between the two projects (I tried putting it in Utils, then in Compiler.Common, then in either Compile or Build, but nothing worked due to various arbitrary reasons). For now I chose to just duplicate the code until I figure out a better way to share command line interfaces between verbs.

I plan on updating the readme once I've made more changes to Build (safety checks, actual incrementality, etc)

### CODE REVIEW

File: src/Microsoft.DotNet.Cli.Utils/ProjectContextExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    internal static class ProjectContextExtensions
+    {
+        public static string GetOutputPath(this ProjectContext context, string configuration, string currentOutputPath)
+        {
+            var outputPath = string.Empty;
+
+            if (string.IsNullOrEmpty(currentOutputPath))
+            {
+                outputPath = Path.Combine(
+                    GetDefaultRootOutputPath(context, currentOutputPath),
```
Comment (dannyvv) [inline src/Microsoft.DotNet.Cli.Utils/ProjectContextExtensions.cs]: I know this is moved code and it was like this, so no need to fix in this commit.
Just want to call out funny behavior where this expression always returns the empty string, so technically it can be removed.

File: src/Microsoft.DotNet.Tools.Builder/Program.cs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            bool isCppMode)
+        {
+            foreach (var context in contexts)
+            {
+                // Set up Output Paths
+                string outputPath = context.GetOutputPath(configValue, outputValue);
+                string intermediateOutputPath = context.GetIntermediateOutputPath(configValue, intermediateValue, outputValue);
+
+                Directory.CreateDirectory(outputPath);
+                Directory.CreateDirectory(intermediateOutputPath);
```
Comment (dannyvv) [inline src/Microsoft.DotNet.Tools.Builder/Program.cs]: Not a comment on this commit, so no need to fix anything since I know this is a direct copy from Compile. 
But CreateDirectory can throw IOException and UnauthorizedAccess exception. Not sure what the patter for error handling in the CLI is here. Just blow up and hope the original exception has enough info, or should we provide custom handling?

**Comment by dannyvv** [comment]
:+1: lgtm

**Comment by davidfowl** [comment]
The boilerplate in these tools are are begging to get inconsistent. The setup of compiler arts etc. we need to fix this before it gets out of hand

**Comment by cdmihai** [comment]
I removed the duplicated code by making the class public and hiding all the public APIs that reference the CommandLineApplication (which is internal).

**Comment by cdmihai** [comment]
@davidfowl I am encapsulating the args in CompileCommandLineApp. I just removed its duplication with the previous commit to this PR :)

For the next PR that includes incremental safety checking, I am going one step further and replacing all the long parameter lists in both Compile and Build with the above mentioned class (slightly modified with maybe other useful things, like reconstructing the argument string).

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Comment by cdmihai** [comment]
@gkhanna79
@schellap

**Comment by cdmihai** [comment]
@dotnet-bot test this please

File: src/Microsoft.DotNet.Tools.Builder/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+dotnet-build -- Orchestrates the compilation of a project and all its dependencies.
+
+**SYNOPSIS**
+dotnet build[options]
+
+**DESCRIPTION**
+
+The build verb orchestrates the compilation of a project: it gathers the dependencies of a project and decides which to compile. 
+
+Users should invoke the Build verb when they want the entire dependency graph compiled, and Compile when they want only a specific project compiled.
```
Comment (gkhanna79) [inline src/Microsoft.DotNet.Tools.Builder/README.md]: Will incremental support be available when "dotnet compile" is used to compile a single project?

File: src/Microsoft.DotNet.Tools.Builder/Program.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+                //compile dependencies
+                if (!CompileDependencies(context, configValue, outputPath, intermediateOutputPath, noHost))
+                {
+                    return false;
+                }
+                
+                //compile project
+                if (!CompileProject(context, configValue, outputPath, intermediateOutputPath, noHost, isNative,
```
Comment (gkhanna79) [inline src/Microsoft.DotNet.Tools.Builder/Program.cs]: Can you please help elaborate on why do we compile the dependencies of the project and then compile project again from the root (that contains the dependencies compiled earlier)?

File: src/Microsoft.DotNet.Tools.Builder/Program.cs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+                .ForwardStdErr()
+                .Execute();
+
+            return compileResult.ExitCode == 0;
+        }
+
+        private static bool CompileProject(ProjectContext context, string configValue, string outputPath,
+            string intermediateOutputPath, bool isNoHost, bool isNative, bool isCppMode, string archValue, string ilcArgsValue,
+            string ilcPathValue, string ilcSdkPathValue, string appDepSdkPath)
+        {
```
Comment (gkhanna79) [inline src/Microsoft.DotNet.Tools.Builder/Program.cs]: How does CompileProject determine which binaries will constitute incremental compilation? I think it is imperative to understand this dataflow since we will need to have that information flow down to the native compilation step as well.

**Comment by gkhanna79** [comment]
Please squash the commits before commit.

File: src/Microsoft.DotNet.Tools.Builder/Program.cs (hunk 5 above)
Comment (schellap) [inline src/Microsoft.DotNet.Tools.Builder/Program.cs]: Could you help us understand how source change affects incremental compilation outputs? i.e., if one line changed in a cs file how do you determine which modules and correspondingly which objs got affected in the native case?

I think CSC will be out of the picture for files that didn't change, how does it work for dotnet compile --native which produces intermediate objs for each IL dll and then these get linked finally to produce exes.

Effectively, how do you compute, only one obj has changed and compiling only that IL dll file and the linking step is required for incremental.

File: src/Microsoft.DotNet.Tools.Builder/README.md (hunk 3 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Builder/README.md]: Incremental compilation will not be available for "dotnet compile". Users have to use Build if they want to build all dependencies and have incrementality.

File: src/Microsoft.DotNet.Tools.Builder/Program.cs (hunk 4 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Builder/Program.cs]: With this PR, the Compile verb no longer compiles its dependencies anymore. Build takes the responsibility of finding out the dependencies and deciding when / if to compile them.

File: src/Microsoft.DotNet.Tools.Builder/Program.cs (hunk 5 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Builder/Program.cs]: @gkhanna79 @schellap 

Incremental compilation will be time stamp based and the granularity is an entire project
Before invoking the Compile verb, the Build verb computes all the input and output files for that project. If all output files are older than all input files, we skip the compilation of the project.

As discussed in the 2 design meetings we've had with the CLI team, for the January and February deadlines we'll do a shared library that both Compile and Build use to compute inputs and outputs. Post Jan / Feb we want to come up with a more uniform way verbs declare their inputs and outputs.

Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Builder/Program.cs]: @gkhanna79 @schellap

Regarding native compilation, I am currently focusing on fleshing out the code for the non-native compilation.

My assumption is that if the bytecode does not need to change (code + satellite resources) then there are no changes for native as well. Is this is a valid assumption? 
If it's not, I'll update the shared IO library to include the native requirements so Build does not know compilation details (e.g., native vs non native)

**Comment by cdmihai** [comment]
@gkhanna79 
I'll squash the commits before merging to master into one large commit (I can also do 3-4 implementation history commits).

I will also cherry pick 89a438acf3e337a628dbf7482434077019f86a3d into this PR, since it belongs here (I only got the refactoring idea while working on #625).

File: src/Microsoft.DotNet.Tools.Builder/README.md (hunk 3 above)
Comment (gkhanna79) [inline src/Microsoft.DotNet.Tools.Builder/README.md]: Can you please help illustrate the difference between dotnet compile and build?

Comment (cdmihai) [inline src/Microsoft.DotNet.Tools.Builder/README.md]: Issue #532 shows the proposed user experience with the Build verb and explains the difference in responsibilities.

Scenarios for using each verb:

The user wants to build a project and all its dependencies. If the incremental safety checks are passed, then the build process is incremental as well. 
The user types: `dotnet build path/to/project`

The user wants to compile only a specific project, without its dependencies, and without incrementality. The user types 'dotnet compile path/to/project'
If the user wants to compile the entire dependency graph with the Compile verb, he needs to call Compile on each dependency.

In summary, the Build verb takes over the Compile verb and users should use Build instead of Compile (unless they want to target specific projects with Compile).

**Comment by cdmihai** [comment]
Since #625 includes the changes in this PR, I'll keep that PR alive and close this one.

**Comment by piotrpMSFT** [comment]
:)
